### PR TITLE
Implement get_inclusion_state serialization

### LINKED ISSUE #871: cclient/serializer/json:get_inclusion_state - serialize response & deserialize reques
This needs implementation, touches:

`cclient/serializer/serializer.h`
`cclient/serializer/json/json_serializer.c`
`cclient/serializer/json/get_inclusion_state.{c,h}`
`cclient/serializer/json/tests/get_inclusion_state.c`

### PR DESCRIPTION
**Pull request by jkrvivian**
- Serialize response
- Deserialize request
- Implement helper function utarray_to_json_boolean_array

Close #871

### CODE REVIEW

File: cclient/serialization/json/helpers.c [hunk 1]
```diff
@@ -99,6 +99,27 @@ retcode_t json_boolean_array_to_utarray(cJSON const* const obj, char const* cons
...
+  if (utarray_len(ut) > 0) {
+    cJSON* array_obj = cJSON_CreateArray();
+    if (array_obj == NULL) {
+      log_critical(json_logger_id, "[%s:%d] %s\n", __func__, __LINE__, STR_CCLIENT_JSON_CREATE);
+      return RC_CCLIENT_JSON_CREATE;
+    }
+
+    cJSON_AddItemToObject(json_root, obj_name, array_obj);
+    int** p = NULL;
+    while ((p = (int*)utarray_next(ut, p))) {
```
Comment (oopsmonk) [inline cclient/serialization/json/helpers.c:116]: incompatible type assignment.

File: cclient/serialization/json/get_inclusion_states.c [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #include "cclient/serialization/json/logger.h"
 
 retcode_t json_get_inclusion_states_serialize_request(const serializer_t *const s,
-                                                      get_inclusion_states_req_t *const obj, char_buffer_t *out) {
+                                                      get_inclusion_states_req_t *const req, char_buffer_t *out) {
   retcode_t ret = RC_OK;
   const char *json_text = NULL;
   size_t len = 0;
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.c]: unused variable.

File: cclient/serialization/json/get_inclusion_states.c [hunk 3]
```diff
@@ -10,7 +10,7 @@
 #include "cclient/serialization/json/logger.h"
 
 retcode_t json_get_inclusion_states_serialize_request(const serializer_t *const s,
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.c]: please correct the const placement as `type const * const variable`

File: cclient/serialization/json/get_inclusion_states.h [hunk 4]
```diff
@@ -20,6 +20,10 @@ extern "C" {
 
 retcode_t json_get_inclusion_states_serialize_request(const serializer_t* const s,
                                                       get_inclusion_states_req_t* const obj, char_buffer_t* out);
+retcode_t json_get_inclusion_states_deserialize_request(const serializer_t* const s, const char* const obj,
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.h]: const placement.

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/tests/get_inclusion_states.c [hunk 5]
```diff
@@ -34,7 +34,47 @@ void test_serialize_get_inclusion_states(void) {
   get_inclusion_states_req_free(&get_is);
 }
 
-void test_deserialize_get_inclusion_states(void) {
+void test_get_inclusion_states_deserialize_request(void) {
+  serializer_t serializer;
+  init_json_serializer(&serializer);
+  const char* json_text = "{\"transactions\": [\"" TEST_81_TRYTES_1
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_inclusion_states.c]: const placement.

File: cclient/serialization/json/tests/get_inclusion_states.c [hunk 6]
```diff
@@ -34,7 +34,47 @@ void test_serialize_get_inclusion_states(void) {
...
+  req_transaction = hash243_queue_at(&deserialize_get_is->tips, 0);
+  flex_trits_from_trytes(hash, NUM_TRITS_HASH, (tryte_t*)TEST_81_TRYTES_2, NUM_TRYTES_HASH, NUM_TRYTES_HASH);
+  TEST_ASSERT_EQUAL_MEMORY(hash, req_transaction, FLEX_TRIT_SIZE_243);
+  get_inclusion_states_req_free(&deserialize_get_is);
+}
+
+void test_get_inclusion_states_serialize_response(void) {
+  serializer_t serializer;
+  init_json_serializer(&serializer);
+  const char* json_text = "{\"states\":[true,false]}";
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_inclusion_states.c]: ditto

File: cclient/serialization/json/tests/get_inclusion_states.c [hunk 7]
```diff
@@ -34,7 +34,47 @@ void test_serialize_get_inclusion_states(void) {
...
+  TEST_ASSERT_EQUAL_STRING(json_text, serializer_out->data);
+
+  char_buffer_free(serializer_out);
+  get_inclusion_states_res_free(&get_is);
+}
+
+void test_get_inclusion_states_deserialize_response(void) {
   serializer_t serializer;
   init_json_serializer(&serializer);
   const char* json_text = "{\"states\": [true,false]}";
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_inclusion_states.c]: ditto

File: cclient/serialization/json/get_inclusion_states.c [hunk 8]
```diff
@@ -48,16 +43,66 @@ retcode_t json_get_inclusion_states_serialize_request(const serializer_t *const
...
+done:
+  cJSON_Delete(json_obj);
+  return ret;
+}
+
+retcode_t json_get_inclusion_states_serialize_response(serializer_t const *const s,
+                                                       get_inclusion_states_res_t const *const res,
+                                                       char_buffer_t *out) {
+  retcode_t ret = RC_OK;
+  const char *json_text = NULL;
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.c]: ditto

File: cclient/serialization/json/get_inclusion_states.c [hunk 9]
```diff
@@ -9,11 +9,10 @@
 #include "cclient/serialization/json/helpers.h"
 #include "cclient/serialization/json/logger.h"
 
-retcode_t json_get_inclusion_states_serialize_request(const serializer_t *const s,
-                                                      get_inclusion_states_req_t *const obj, char_buffer_t *out) {
+retcode_t json_get_inclusion_states_serialize_request(serializer_t const *const s,
+                                                      get_inclusion_states_req_t const *const req, char_buffer_t *out) {
   retcode_t ret = RC_OK;
   const char *json_text = NULL;
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.c]: ditto

File: cclient/serialization/json/tests/get_inclusion_states.c [hunk 10]
```diff
@@ -7,7 +7,7 @@
 
 #include "cclient/serialization/json/tests/shared.h"
 
-void test_serialize_get_inclusion_states(void) {
+void test_get_inclusion_states_serialize_request(void) {
   serializer_t serializer;
   init_json_serializer(&serializer);
   const char* json_text =
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_inclusion_states.c]: ditto

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: cclient/serialization/json/get_inclusion_states.c [hunk 11]
```diff
@@ -48,16 +43,66 @@ retcode_t json_get_inclusion_states_serialize_request(const serializer_t *const
...
   log_debug(json_logger_id, "[%s:%d] %s\n", __func__, __LINE__, obj);
   JSON_CHECK_ERROR(json_obj, json_item, json_logger_id);
 
-  ret = json_boolean_array_to_utarray(json_obj, "states", out->states);
+  ret = json_array_to_hash243_queue(json_obj, "transactions", &req->hashes);
+  if (ret != RC_OK) {
+    goto done;
+  }
+
+  ret = json_array_to_hash243_queue(json_obj, "tips", &req->tips);
```
Comment (oopsmonk) [inline cclient/serialization/json/get_inclusion_states.c]: `tips` is optional, please check via `cJSON_HasObjectItem`

File: cclient/serialization/json/tests/get_inclusion_states.c [hunk 12]
```diff
@@ -7,22 +7,22 @@
 
 #include "cclient/serialization/json/tests/shared.h"
 
-void test_serialize_get_inclusion_states(void) {
+void test_get_inclusion_states_serialize_request(void) {
```
Comment (oopsmonk) [inline cclient/serialization/json/tests/get_inclusion_states.c:10]: please add test cases with no `tips`

**Review by oopsmonk — CHANGES_REQUESTED** [review]

**Comment by thibault-martinez** [comment]
Hi @jkrvivian , could you check Sam's latest comments ? I'm going to need this serialization soon :) Thanks in advance !

**Comment by jkrvivian** [comment]
Sorry for late update, I've modified it and pushed it!

**Review by oopsmonk — APPROVED** [review]

**Comment by oopsmonk** [comment]
appreciate your help

**Pull request merged**